File: java/client/src/test/java/org/signal/libsignal/net/UnauthKeysServiceTest.kt
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.net

import org.junit.Test
import org.signal.libsignal.internal.TokioAsyncContext
import org.signal.libsignal.protocol.IdentityKey
import org.signal.libsignal.protocol.ServiceId
import org.signal.libsignal.protocol.ecc.ECPublicKey
import org.signal.libsignal.protocol.kem.KEMPublicKey
import org.signal.libsignal.protocol.state.PreKeyBundle
import org.signal.libsignal.zkgroup.groupsend.GroupSendFullToken
import java.util.UUID
import java.util.concurrent.TimeUnit
import kotlin.io.encoding.Base64
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertIs

class UnauthKeysServiceTest {
  companion object {
    private val aci = ServiceId.Aci(UUID.fromString("9d0652a3-dcc3-4d11-975f-74d61598733f"))
    private val pni = ServiceId.Pni(UUID.fromString("000002a3-dcc3-4d11-975f-74d61598733f"))
    private val deviceId = 2
    private val registrationId = 1234
    private val preKeyId = 5
    private val signedPreKeyId = 7
    private val kyberPreKeyId = 9
    private val secondDeviceId = 3
    private val secondRegistrationId = 5678
    private val secondPreKeyId = 11
    private val secondSignedPreKeyId = 13
    private val secondKyberPreKeyId = 15

    // [0x11; 16]
    private val testAccessKey = Base64.decode("EREREREREREREREREREREQ==")
    private val identityKey = dummyIdentityKey(0x12)

    private val signedPreKeyPublic = dummyEcPublicKey(0x34)
    private val signedPreKeySignature = ByteArray(64) { 0x56 }
    private val kyberPreKeyPublic = dummyKemPublicKey(0x78)
    private val kyberPreKeySignature = ByteArray(64) { 0x9A.toByte() }
    private val preKeyPublic = dummyEcPublicKey(0x43)
    private val secondPreKeyPublic = dummyEcPublicKey(0xD4.toByte())

    private val secondSignedPreKeyPublic = dummyEcPublicKey(0x21)
    private val secondSignedPreKeySignature = ByteArray(64) { 0x32 }
    private val secondKyberPreKeyPublic = dummyKemPublicKey(0x64)
    private val secondKyberPreKeySignature = ByteArray(64) { 0x64 }

    private val testGroupSendToken =
      GroupSendFullToken(
        Base64.decode("ABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABo5c+LAQAA"),
      )

    private fun dummyEcPublicKey(fill: Byte): ECPublicKey = ECPublicKey(byteArrayOf(0x05) + ByteArray(32) { fill })

    private fun dummyIdentityKey(fill: Byte): IdentityKey = IdentityKey(dummyEcPublicKey(fill))

    // kyber1024::Parameters::PUBLIC_KEY_LENGTH
    // This needs to be const to avoid java initialization order issues
    private const val KYBER1024_PUBLIC_KEY_LENGTH: Int = 1568

    private fun dummyKemPublicKey(fill: Byte): KEMPublicKey =
      KEMPublicKey(
        byteArrayOf(0x08) +
          ByteArray(KYBER1024_PUBLIC_KEY_LENGTH) { fill },
      )
  }

  private fun doTestSingleKeyWithPreKey(
    specifierString: String,
    specifier: DeviceSpecifier,
    authHeaders: Map<String, String>,
    authValue: UserBasedAuthorization,
  ) {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )
    val service = UnauthKeysService(chat)
    val responseFuture =
      service.getPreKeys(
        aci,
        specifier,
        authValue,
      )
    // If the future is done immediately, then there's some internal error.
    if (responseFuture.isDone) {
      throw AssertionError(responseFuture.get().toString())
    }
    val (request, requestId) = fakeRemote.nextIncomingRequest.get(1, TimeUnit.SECONDS)
    assertEquals("GET", request.method)
    assertEquals("/v2/keys/${aci.rawUUID}/$specifierString", request.pathAndQuery)
    assertEquals(authHeaders, request.headers)
    assertEquals(0, request.body.size)
    val responseBody = """
           {
               "identityKey": "${Base64.encode(identityKey.serialize())}",
               "devices": [{
                   "deviceId": $deviceId,
                   "registrationId": $registrationId,
                   "preKey": {
                       "keyId": $preKeyId,
                       "publicKey": "${Base64.encode(preKeyPublic.serialize())}"
                   },
                   "signedPreKey": {
                       "keyId": $signedPreKeyId,
                       "publicKey": "${Base64.encode(signedPreKeyPublic.serialize())}",
                       "signature": "${Base64.encode(signedPreKeySignature)}"
                   },
                   "pqPreKey": {
                       "keyId": $kyberPreKeyId,
                       "publicKey": "${Base64.encode(kyberPreKeyPublic.serialize())}",
                       "signature": "${Base64.encode(kyberPreKeySignature)}"
                   }
               }]
           }
          """
    fakeRemote.sendResponse(
      requestId,
      200,
      "OK",
      arrayOf("content-type: application/json"),
      responseBody.encodeToByteArray(),
    )
    val response = responseFuture.get()
    val (rik, bundles) = assertIs<RequestResult.Success<Pair<IdentityKey, List<PreKeyBundle>>>>(response).result
    assertEquals(identityKey, rik)
    assertEquals(1, bundles.size)
    val bundle = bundles[0]
    assertEquals(bundle.kyberPreKeyId, kyberPreKeyId)
    assertEquals(bundle.preKeyId, preKeyId)
    assertEquals(bundle.preKey, preKeyPublic)
    assertEquals(bundle.signedPreKey, signedPreKeyPublic)
    assertContentEquals(bundle.signedPreKeySignature, signedPreKeySignature)
    assertEquals(bundle.kyberPreKey, kyberPreKeyPublic)
    assertContentEquals(bundle.kyberPreKeySignature, kyberPreKeySignature)
  }

  @Test
  fun testSingleKeyWithPreKey_allDevices_accessKey() =
    doTestSingleKeyWithPreKey(
      "*",
      DeviceSpecifier.AllDevices,
      mapOf("unidentified-access-key" to Base64.encode(testAccessKey)),
      UserBasedAuthorization.AccessKey(testAccessKey),
    )

  @Test
  fun testSingleKeyWithPreKey_allDevices_groupSend() =
    doTestSingleKeyWithPreKey(
      "*",
      DeviceSpecifier.AllDevices,
      mapOf(
        "group-send-token" to
          Base64.encode(testGroupSendToken.serialize()),
      ),
      UserBasedAuthorization.GroupSend(testGroupSendToken),
    )

  @Test
  fun testSingleKeyWithPreKey_allDevices_unrestricted() =
    doTestSingleKeyWithPreKey(
      "*",
      DeviceSpecifier.AllDevices,
      mapOf("unidentified-access-key" to Base64.encode(ByteArray(16))),
      UserBasedAuthorization.UnrestrictedUnauthenticatedAccess,
    )

  @Test
  fun testSingleKeyWithPreKey_specificDevice_accessKey() =
    doTestSingleKeyWithPreKey(
      deviceId.toString(),
      DeviceSpecifier.SpecificDevice(deviceId),
      mapOf("unidentified-access-key" to Base64.encode(testAccessKey)),
      UserBasedAuthorization.AccessKey(testAccessKey),
    )

  @Test
  fun testSingleKeyWithPreKey_specificDevice_groupSend() =
    doTestSingleKeyWithPreKey(
      deviceId.toString(),
      DeviceSpecifier.SpecificDevice(deviceId),
      mapOf(
        "group-send-token" to
          Base64.encode(testGroupSendToken.serialize()),
      ),
      UserBasedAuthorization.GroupSend(testGroupSendToken),
    )

  @Test
  fun testSingleKeyWithPreKey_specificDevice_unrestricted() =
    doTestSingleKeyWithPreKey(
      deviceId.toString(),
      DeviceSpecifier.SpecificDevice(deviceId),
      mapOf("unidentified-access-key" to Base64.encode(ByteArray(16))),
      UserBasedAuthorization.UnrestrictedUnauthenticatedAccess,
    )

  @Test
  fun testSingleKeyNoPreKey() {
    for ((specifierString, specifier) in listOf(
      Pair("*", DeviceSpecifier.AllDevices),
      Pair(deviceId.toString(), DeviceSpecifier.SpecificDevice(deviceId)),
    )) {
      for ((authHeaders, authValue) in listOf(
        Pair(
          mapOf("unidentified-access-key" to Base64.encode(testAccessKey)),
          UserBasedAuthorization.AccessKey(testAccessKey),
        ),
        Pair(
          mapOf("group-send-token" to Base64.encode(testGroupSendToken.serialize())),
          UserBasedAuthorization.GroupSend(testGroupSendToken),
        ),
        Pair(
          mapOf("unidentified-access-key" to Base64.encode(ByteArray(16))),
          UserBasedAuthorization.UnrestrictedUnauthenticatedAccess,
        ),
      )) {
        val tokioAsyncContext = TokioAsyncContext()
        val (chat, fakeRemote) =
          UnauthenticatedChatConnection.fakeConnect(
            tokioAsyncContext,
            NoOpListener(),
            Network.Environment.STAGING,
          )
        val service = UnauthKeysService(chat)
        val responseFuture =
          service.getPreKeys(
            aci,
            specifier,
            authValue,
          )
        // If the future is done immediately, then there's some internal error.
        if (responseFuture.isDone) {
          throw AssertionError(responseFuture.get().toString())
        }
        val (request, requestId) = fakeRemote.nextIncomingRequest.get(1, TimeUnit.SECONDS)
        assertEquals("GET", request.method)
        assertEquals("/v2/keys/${aci.rawUUID}/$specifierString", request.pathAndQuery)
        assertEquals(authHeaders, request.headers)
        assertEquals(0, request.body.size)
        val responseBody = """
           {
               "identityKey": "${Base64.encode(identityKey.serialize())}",
               "devices": [{
                   "deviceId": $deviceId,
                   "registrationId": $registrationId,
                   "signedPreKey": {
                       "keyId": $signedPreKeyId,
                       "publicKey": "${Base64.encode(signedPreKeyPublic.serialize())}",
                       "signature": "${Base64.encode(signedPreKeySignature)}"
                   },
                   "pqPreKey": {
                       "keyId": $kyberPreKeyId,
                       "publicKey": "${Base64.encode(kyberPreKeyPublic.serialize())}",
                       "signature": "${Base64.encode(kyberPreKeySignature)}"
                   }
               }]
           }
          """
        fakeRemote.sendResponse(
          requestId,
          200,
          "OK",
          arrayOf("content-type: application/json"),
          responseBody.encodeToByteArray(),
        )
        val response = responseFuture.get()
        val (rik, bundles) = assertIs<RequestResult.Success<Pair<IdentityKey, List<PreKeyBundle>>>>(response).result
        assertEquals(identityKey, rik)
        assertEquals(1, bundles.size)
        val bundle = bundles[0]
        assertEquals(bundle.kyberPreKeyId, kyberPreKeyId)
        assertEquals(bundle.preKeyId, -1)
        assertEquals(bundle.preKey, null)
        assertEquals(bundle.signedPreKey, signedPreKeyPublic)
        assertContentEquals(bundle.signedPreKeySignature, signedPreKeySignature)
        assertEquals(bundle.kyberPreKey, kyberPreKeyPublic)
        assertContentEquals(bundle.kyberPreKeySignature, kyberPreKeySignature)
      }
    }
  }

  @Test
  fun testAllKeys() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )
    val service = UnauthKeysService(chat)
    val responseFuture =
      service.getPreKeys(
        pni,
        DeviceSpecifier.AllDevices,
        UserBasedAuthorization.AccessKey(testAccessKey),
      )
    // If the future is done immediately, then there's some internal error.
    if (responseFuture.isDone) {
      throw AssertionError(responseFuture.get().toString())
    }
    val (request, requestId) = fakeRemote.nextIncomingRequest.get(1, TimeUnit.SECONDS)
    assertEquals("GET", request.method)
    assertEquals("/v2/keys/${pni.toServiceIdString()}/*", request.pathAndQuery)
    assertEquals(
      mapOf(
        "unidentified-access-key" to Base64.encode(testAccessKey),
      ),
      request.headers,
    )
    assertEquals(0, request.body.size)
    val responseBody = """
          {
              "identityKey": "${Base64.encode(identityKey.serialize())}",
              "devices": [
                  {
                      "deviceId": $deviceId,
                      "registrationId": $registrationId,
                      "signedPreKey": {
                          "keyId": $signedPreKeyId,
                          "publicKey": "${Base64.encode(signedPreKeyPublic.serialize())}",
                          "signature": "${Base64.encode(signedPreKeySignature)}"
                      },
                      "preKey": {
                          "keyId": $preKeyId,
                          "publicKey": "${Base64.encode(preKeyPublic.serialize())}"
                      },
                      "pqPreKey": {
                          "keyId": $kyberPreKeyId,
                          "publicKey": "${Base64.encode(kyberPreKeyPublic.serialize())}",
                          "signature": "${Base64.encode(kyberPreKeySignature)}"
                      }
                  },
                  {
                      "deviceId": $secondDeviceId,
                      "registrationId": $secondRegistrationId,
                      "signedPreKey": {
                          "keyId": $secondSignedPreKeyId,
                          "publicKey": "${Base64.encode(secondSignedPreKeyPublic.serialize())}",
                          "signature": "${Base64.encode(secondSignedPreKeySignature)}"
                      },
                      "preKey": {
                          "keyId": $secondPreKeyId,
                          "publicKey": "${Base64.encode(secondPreKeyPublic.serialize())}"
                      },
                      "pqPreKey": {
                          "keyId": $secondKyberPreKeyId,
                          "publicKey": "${Base64.encode(secondKyberPreKeyPublic.serialize())}",
                          "signature": "${Base64.encode(secondKyberPreKeySignature)}"
                      }
                  }
              ]
          }
          """
    fakeRemote.sendResponse(
      requestId,
      200,
      "OK",
      arrayOf("content-type: application/json"),
      responseBody.encodeToByteArray(),
    )
    val response = responseFuture.get()
    val (rik, bundles) = assertIs<RequestResult.Success<Pair<IdentityKey, List<PreKeyBundle>>>>(response).result
    assertEquals(identityKey, rik)
    assertEquals(2, bundles.size)
    assertEquals(bundles[0].kyberPreKeyId, kyberPreKeyId)
    assertEquals(bundles[0].preKeyId, preKeyId)
    assertEquals(bundles[0].preKey, preKeyPublic)
    assertEquals(bundles[0].signedPreKey, signedPreKeyPublic)
    assertContentEquals(bundles[0].signedPreKeySignature, signedPreKeySignature)
    assertEquals(bundles[0].kyberPreKey, kyberPreKeyPublic)
    assertContentEquals(bundles[0].kyberPreKeySignature, kyberPreKeySignature)
    assertEquals(bundles[1].kyberPreKeyId, secondKyberPreKeyId)
    assertEquals(bundles[1].preKeyId, secondPreKeyId)
    assertEquals(bundles[1].preKey, secondPreKeyPublic)
    assertEquals(bundles[1].signedPreKey, secondSignedPreKeyPublic)
    assertContentEquals(bundles[1].signedPreKeySignature, secondSignedPreKeySignature)
    assertEquals(bundles[1].kyberPreKey, secondKyberPreKeyPublic)
    assertContentEquals(bundles[1].kyberPreKeySignature, secondKyberPreKeySignature)
  }

  @Test
  fun testUnauthorized() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )
    val service = UnauthKeysService(chat)
    val responseFuture =
      service.getPreKeys(
        aci,
        DeviceSpecifier.AllDevices,
        UserBasedAuthorization.AccessKey(testAccessKey),
      )
    val (request, requestId) = fakeRemote.nextIncomingRequest.get(1, TimeUnit.SECONDS)
    assertEquals("GET", request.method)
    assertEquals("/v2/keys/${aci.rawUUID}/*", request.pathAndQuery)
    assertEquals(
      mapOf(
        "unidentified-access-key" to Base64.encode(testAccessKey),
      ),
      request.headers,
    )
    assertEquals(0, request.body.size)
    fakeRemote.sendResponse(requestId, 401, "Unauthorized", arrayOf(), byteArrayOf())
    val error = assertIs<RequestResult.NonSuccess<GetPreKeysError>>(responseFuture.get()).error
    assertIs<RequestUnauthorizedException>(error)
  }

  @Test
  fun testNotFound() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )
    val service = UnauthKeysService(chat)
    val responseFuture =
      service.getPreKeys(
        aci,
        DeviceSpecifier.AllDevices,
        UserBasedAuthorization.AccessKey(testAccessKey),
      )
    val (request, requestId) = fakeRemote.nextIncomingRequest.get(1, TimeUnit.SECONDS)
    assertEquals("GET", request.method)
    assertEquals("/v2/keys/${aci.rawUUID}/*", request.pathAndQuery)
    assertEquals(
      mapOf(
        "unidentified-access-key" to Base64.encode(testAccessKey),
      ),
      request.headers,
    )
    assertEquals(0, request.body.size)
    fakeRemote.sendResponse(requestId, 404, "Not Found", arrayOf(), byteArrayOf())
    val error = assertIs<RequestResult.NonSuccess<GetPreKeysError>>(responseFuture.get()).error
    assertIs<ServiceIdNotFoundException>(error)
  }
}

File: node/ts/test/chat/UnauthKeysServiceTest.ts
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { config, expect, use } from 'chai';
import chaiAsPromised from 'chai-as-promised';

import * as Native from '../../Native.js';
import * as util from '../util.js';
import { TokioAsyncContext, UnauthKeysService } from '../../net.js';
import { connectUnauth } from './ServiceTestUtils.js';
import { Aci } from '../../Address.js';
import { GroupSendFullToken } from '../../zkgroup/index.js';
import { PublicKey } from '../../EcKeys.js';
import { ErrorCode, KEMPublicKey, LibSignalErrorBase } from '../../index.js';
import { fromBase64, toBase64 } from '../util.js';

use(chaiAsPromised);

util.initLogger();
config.truncateThreshold = 0;

describe('UnauthKeysService', () => {
  describe('getPreKeys', () => {
    const ACI = Aci.fromUuid('9d0652a3-dcc3-4d11-975f-74d61598733f');
    const DEVICE_ID = 2;
    const REGISTRATION_ID = 1234;
    const PRE_KEY_ID = 5;
    const SIGNED_PRE_KEY_ID = 7;
    const KYBER_PRE_KEY_ID = 9;
    const SECOND_DEVICE_ID = 3;
    const SECOND_REGISTRATION_ID = 5678;
    const SECOND_PRE_KEY_ID = 11;
    const SECOND_SIGNED_PRE_KEY_ID = 13;
    const SECOND_KYBER_PRE_KEY_ID = 15;

    // [0x11; 16]
    const TEST_ACCESS_KEY = fromBase64('EREREREREREREREREREREQ==');
    const IDENTITY_KEY = dummyEcPublicKey(0x12);

    const SIGNED_PRE_KEY_PUBLIC = dummyEcPublicKey(0x34);
    const SIGNED_PRE_KEY_SIGNATURE = repeatedBytes(0x56, 64);
    const KYBER_PRE_KEY_PUBLIC = dummyKemPublicKey(0x78);
    const KYBER_PRE_KEY_SIGNATURE = repeatedBytes(0x9a, 64);
    const PRE_KEY_PUBLIC = dummyEcPublicKey(0x43);

    const SECOND_PRE_KEY_PUBLIC = dummyEcPublicKey(0xd4);
    const SECOND_SIGNED_PRE_KEY_PUBLIC = dummyEcPublicKey(0x21);
    const SECOND_SIGNED_PRE_KEY_SIGNATURE = repeatedBytes(0x32, 64);
    const SECOND_KYBER_PRE_KEY_PUBLIC = dummyKemPublicKey(0x64);
    const SECOND_KYBER_PRE_KEY_SIGNATURE = repeatedBytes(0x64, 64);

    const TEST_GROUP_SEND_TOKEN = new GroupSendFullToken(
      fromBase64('ABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABo5c+LAQAA')
    );

    function repeatedBytes(
      fill: number,
      count: number
    ): Uint8Array<ArrayBuffer> {
      const out = new Uint8Array<ArrayBuffer>(new ArrayBuffer(count));
      for (let i = 0; i < count; i++) {
        out[i] = fill;
      }
      return out;
    }

    function bytePrefix(
      prefixByte: number,
      buf: Uint8Array<ArrayBuffer>
    ): Uint8Array<ArrayBuffer> {
      const out = new Uint8Array<ArrayBuffer>(new ArrayBuffer(1 + buf.length));
      out.set([prefixByte], 0);
      out.set(buf, 1);
      return out;
    }

    function dummyEcPublicKey(fill: number): PublicKey {
      return PublicKey.deserialize(bytePrefix(0x05, repeatedBytes(fill, 32)));
    }

    function dummyKemPublicKey(fill: number): KEMPublicKey {
      return KEMPublicKey.deserialize(
        // 1568 is kyber1024::Parameters::PUBLIC_KEY_LENGTH
        bytePrefix(0x08, repeatedBytes(fill, 1568))
      );
    }

    it('test single key with prekey', async () => {
      for (const { specifierString, specifier } of [
        { specifierString: '*', specifier: 'all' as const },
        { specifierString: `${DEVICE_ID}`, specifier: { deviceId: DEVICE_ID } },
      ]) {
        for (const { authHeaders, authValue } of [
          {
            authHeaders: new Map([
              ['unidentified-access-key', toBase64(TEST_ACCESS_KEY)],
            ]),
            authValue: { accessKey: TEST_ACCESS_KEY },
          },
          {
            authHeaders: new Map([
              ['group-send-token', toBase64(TEST_GROUP_SEND_TOKEN.serialize())],
            ]),
            authValue: TEST_GROUP_SEND_TOKEN,
          },
        ]) {
          const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
          const [service, fakeRemote] = connectUnauth<UnauthKeysService>(tokio);
          const responseFuture = service.getPreKeys({
            auth: authValue,
            target: ACI,
            device: specifier,
          });
          const request = await fakeRemote.assertReceiveIncomingRequest();
          expect(request.verb).to.eq('GET');
          expect(request.path).to.eq(
            `/v2/keys/${ACI.getRawUuid()}/${specifierString}`
          );
          expect(request.headers).to.deep.eq(authHeaders);
          expect(request.body.length).to.eq(0);
          fakeRemote.sendReplyTo(request, {
            status: 200,
            message: 'OK',
            headers: ['content-type: application/json'],
            body: new TextEncoder().encode(
              JSON.stringify({
                identityKey: toBase64(IDENTITY_KEY.serialize()),
                devices: [
                  {
                    deviceId: DEVICE_ID,
                    registrationId: REGISTRATION_ID,
                    preKey: {
                      keyId: PRE_KEY_ID,
                      publicKey: toBase64(PRE_KEY_PUBLIC.serialize()),
                    },
                    signedPreKey: {
                      keyId: SIGNED_PRE_KEY_ID,
                      publicKey: toBase64(SIGNED_PRE_KEY_PUBLIC.serialize()),
                      signature: toBase64(SIGNED_PRE_KEY_SIGNATURE),
                    },
                    pqPreKey: {
                      keyId: KYBER_PRE_KEY_ID,
                      publicKey: toBase64(KYBER_PRE_KEY_PUBLIC.serialize()),
                      signature: toBase64(KYBER_PRE_KEY_SIGNATURE),
                    },
                  },
                ],
              })
            ),
          });
          const { identityKey, preKeyBundles } = await responseFuture;
          expect(identityKey.serialize()).to.deep.eq(IDENTITY_KEY.serialize());
          expect(preKeyBundles.length).to.eq(1);
          const bundle = preKeyBundles[0];
          expect(bundle.kyberPreKeyId()).to.eq(KYBER_PRE_KEY_ID);
          expect(bundle.preKeyId()).to.eq(PRE_KEY_ID);
          expect(bundle.preKeyPublic()?.serialize()).to.deep.eq(
            PRE_KEY_PUBLIC.serialize()
          );
          expect(bundle.signedPreKeyPublic().serialize()).to.deep.eq(
            SIGNED_PRE_KEY_PUBLIC.serialize()
          );
          expect(bundle.signedPreKeySignature()).to.deep.eq(
            SIGNED_PRE_KEY_SIGNATURE
          );
          expect(bundle.kyberPreKeyPublic().serialize()).to.deep.eq(
            KYBER_PRE_KEY_PUBLIC.serialize()
          );
          expect(bundle.kyberPreKeySignature()).to.deep.eq(
            KYBER_PRE_KEY_SIGNATURE
          );
        }
      }
    });
    it('test single key with no prekey', async () => {
      for (const { specifierString, specifier } of [
        { specifierString: '*', specifier: 'all' as const },
        { specifierString: `${DEVICE_ID}`, specifier: { deviceId: DEVICE_ID } },
      ]) {
        for (const { authHeaders, authValue } of [
          {
            authHeaders: new Map([
              ['unidentified-access-key', toBase64(TEST_ACCESS_KEY)],
            ]),
            authValue: { accessKey: TEST_ACCESS_KEY },
          },
          {
            authHeaders: new Map([
              ['group-send-token', toBase64(TEST_GROUP_SEND_TOKEN.serialize())],
            ]),
            authValue: TEST_GROUP_SEND_TOKEN,
          },
        ]) {
          const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
          const [service, fakeRemote] = connectUnauth<UnauthKeysService>(tokio);
          const responseFuture = service.getPreKeys({
            auth: authValue,
            target: ACI,
            device: specifier,
          });
          const request = await fakeRemote.assertReceiveIncomingRequest();
          expect(request.verb).to.eq('GET');
          expect(request.path).to.eq(
            `/v2/keys/${ACI.getRawUuid()}/${specifierString}`
          );
          expect(request.headers).to.deep.eq(authHeaders);
          expect(request.body.length).to.eq(0);
          fakeRemote.sendReplyTo(request, {
            status: 200,
            message: 'OK',
            headers: ['content-type: application/json'],
            body: new TextEncoder().encode(
              JSON.stringify({
                identityKey: toBase64(IDENTITY_KEY.serialize()),
                devices: [
                  {
                    deviceId: DEVICE_ID,
                    registrationId: REGISTRATION_ID,
                    signedPreKey: {
                      keyId: SIGNED_PRE_KEY_ID,
                      publicKey: toBase64(SIGNED_PRE_KEY_PUBLIC.serialize()),
                      signature: toBase64(SIGNED_PRE_KEY_SIGNATURE),
                    },
                    pqPreKey: {
                      keyId: KYBER_PRE_KEY_ID,
                      publicKey: toBase64(KYBER_PRE_KEY_PUBLIC.serialize()),
                      signature: toBase64(KYBER_PRE_KEY_SIGNATURE),
                    },
                  },
                ],
              })
            ),
          });
          const { identityKey, preKeyBundles } = await responseFuture;
          expect(identityKey.serialize()).to.deep.eq(IDENTITY_KEY.serialize());
          expect(preKeyBundles.length).to.eq(1);
          const bundle = preKeyBundles[0];
          expect(bundle.kyberPreKeyId()).to.eq(KYBER_PRE_KEY_ID);
          expect(bundle.preKeyId()).to.eq(null);
          expect(bundle.preKeyPublic()).to.eq(null);
          expect(bundle.signedPreKeyPublic().serialize()).to.deep.eq(
            SIGNED_PRE_KEY_PUBLIC.serialize()
          );
          expect(bundle.signedPreKeySignature()).to.deep.eq(
            SIGNED_PRE_KEY_SIGNATURE
          );
          expect(bundle.kyberPreKeyPublic().serialize()).to.deep.eq(
            KYBER_PRE_KEY_PUBLIC.serialize()
          );
          expect(bundle.kyberPreKeySignature()).to.deep.eq(
            KYBER_PRE_KEY_SIGNATURE
          );
        }
      }
    });
    it('test all keys', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [service, fakeRemote] = connectUnauth<UnauthKeysService>(tokio);
      const responseFuture = service.getPreKeys({
        auth: { accessKey: TEST_ACCESS_KEY },
        target: ACI,
        device: 'all',
      });
      const request = await fakeRemote.assertReceiveIncomingRequest();
      expect(request.verb).to.eq('GET');
      expect(request.path).to.eq(`/v2/keys/${ACI.getRawUuid()}/*`);
      expect(request.headers).to.deep.eq(
        new Map([['unidentified-access-key', toBase64(TEST_ACCESS_KEY)]])
      );
      expect(request.body.length).to.eq(0);
      fakeRemote.sendReplyTo(request, {
        status: 200,
        message: 'OK',
        headers: ['content-type: application/json'],
        body: new TextEncoder().encode(
          JSON.stringify({
            identityKey: toBase64(IDENTITY_KEY.serialize()),
            devices: [
              {
                deviceId: DEVICE_ID,
                registrationId: REGISTRATION_ID,
                signedPreKey: {
                  keyId: SIGNED_PRE_KEY_ID,
                  publicKey: toBase64(SIGNED_PRE_KEY_PUBLIC.serialize()),
                  signature: toBase64(SIGNED_PRE_KEY_SIGNATURE),
                },
                preKey: {
                  keyId: PRE_KEY_ID,
                  publicKey: toBase64(PRE_KEY_PUBLIC.serialize()),
                },
                pqPreKey: {
                  keyId: KYBER_PRE_KEY_ID,
                  publicKey: toBase64(KYBER_PRE_KEY_PUBLIC.serialize()),
                  signature: toBase64(KYBER_PRE_KEY_SIGNATURE),
                },
              },
              {
                deviceId: SECOND_DEVICE_ID,
                registrationId: SECOND_REGISTRATION_ID,
                signedPreKey: {
                  keyId: SECOND_SIGNED_PRE_KEY_ID,
                  publicKey: toBase64(SECOND_SIGNED_PRE_KEY_PUBLIC.serialize()),
                  signature: toBase64(SECOND_SIGNED_PRE_KEY_SIGNATURE),
                },
                preKey: {
                  keyId: SECOND_PRE_KEY_ID,
                  publicKey: toBase64(SECOND_PRE_KEY_PUBLIC.serialize()),
                },
                pqPreKey: {
                  keyId: SECOND_KYBER_PRE_KEY_ID,
                  publicKey: toBase64(SECOND_KYBER_PRE_KEY_PUBLIC.serialize()),
                  signature: toBase64(SECOND_KYBER_PRE_KEY_SIGNATURE),
                },
              },
            ],
          })
        ),
      });
      const { identityKey, preKeyBundles } = await responseFuture;
      expect(identityKey.serialize()).to.deep.eq(IDENTITY_KEY.serialize());
      expect(preKeyBundles.length).to.eq(2);
      expect(preKeyBundles[0].kyberPreKeyId()).to.eq(KYBER_PRE_KEY_ID);
      expect(preKeyBundles[0].preKeyId()).to.eq(PRE_KEY_ID);
      expect(preKeyBundles[0].preKeyPublic()?.serialize()).to.deep.eq(
        PRE_KEY_PUBLIC.serialize()
      );
      expect(preKeyBundles[0].signedPreKeyPublic().serialize()).to.deep.eq(
        SIGNED_PRE_KEY_PUBLIC.serialize()
      );
      expect(preKeyBundles[0].signedPreKeySignature()).to.deep.eq(
        SIGNED_PRE_KEY_SIGNATURE
      );
      expect(preKeyBundles[0].kyberPreKeyPublic().serialize()).to.deep.eq(
        KYBER_PRE_KEY_PUBLIC.serialize()
      );
      expect(preKeyBundles[0].kyberPreKeySignature()).to.deep.eq(
        KYBER_PRE_KEY_SIGNATURE
      );
      expect(preKeyBundles[1].kyberPreKeyId()).to.eq(SECOND_KYBER_PRE_KEY_ID);
      expect(preKeyBundles[1].preKeyId()).to.eq(SECOND_PRE_KEY_ID);
      expect(preKeyBundles[1].preKeyPublic()?.serialize()).to.deep.eq(
        SECOND_PRE_KEY_PUBLIC.serialize()
      );
      expect(preKeyBundles[1].signedPreKeyPublic().serialize()).to.deep.eq(
        SECOND_SIGNED_PRE_KEY_PUBLIC.serialize()
      );
      expect(preKeyBundles[1].signedPreKeySignature()).to.deep.eq(
        SECOND_SIGNED_PRE_KEY_SIGNATURE
      );
      expect(preKeyBundles[1].kyberPreKeyPublic().serialize()).to.deep.eq(
        SECOND_KYBER_PRE_KEY_PUBLIC.serialize()
      );
      expect(preKeyBundles[1].kyberPreKeySignature()).to.deep.eq(
        SECOND_KYBER_PRE_KEY_SIGNATURE
      );
    });
    it('should properly handle unauthorized', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthKeysService>(tokio);
      const responseFuture = chat.getPreKeys({
        auth: TEST_GROUP_SEND_TOKEN,
        target: ACI,
        device: 'all',
      });
      const request = await fakeRemote.assertReceiveIncomingRequest();
      fakeRemote.sendReplyTo(request, {
        status: 401,
        message: 'Unauthorized',
      });
      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.RequestUnauthorized,
        });
    });
    it('should properly handle not found', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthKeysService>(tokio);
      const responseFuture = chat.getPreKeys({
        auth: TEST_GROUP_SEND_TOKEN,
        target: ACI,
        device: 'all',
      });
      const request = await fakeRemote.assertReceiveIncomingRequest();
      fakeRemote.sendReplyTo(request, {
        status: 404,
        message: 'Not Found',
      });
      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.ServiceIdNotFound,
        });
    });
  });
});
